File: py/stats.py
"""
stats.py - 流量统计引擎
维护滑动窗口，计算 Cur / Avg / Min / Max / Ttl 五项指标。
"""

from collections import deque
from dataclasses import dataclass
from typing import Optional

from collector import Snapshot


@dataclass
class TrafficStats:
    """某一方向（收/发）的统计结果"""
    current: float = 0.0     # 当前速率 (bytes/s)
    average: float = 0.0     # 平均速率 (bytes/s)
    minimum: float = float("inf")   # 最小速率 (bytes/s)
    maximum: float = 0.0     # 最大速率 (bytes/s)
    total: int = 0           # 累计字节数


class StatisticsEngine:
    """
    为单个网卡维护收发两个方向的统计数据。
    滑动窗口保存历史快照，据此计算各项指标。
    """

    def __init__(
        self,
        refresh_interval_ms: int = 500,
        average_window_sec: int = 300,
    ):
        self._refresh_ms = refresh_interval_ms
        self._avg_window_sec = average_window_sec

        # 滑动窗口
        max_samples = max(
            int(1000 / refresh_interval_ms * average_window_sec), 600
        )
        self._samples: deque[Snapshot] = deque(maxlen=max_samples)

        # 计算 1 秒窗口需要的样本数
        self._second_window = max(int(1000 / refresh_interval_ms), 1)

        # 统计结果
        self.incoming = TrafficStats()
        self.outgoing = TrafficStats()

        # 用于记录本次速率的历史（供 graph 使用）
        self._incoming_history: deque[float] = deque(maxlen=1024)
        self._outgoing_history: deque[float] = deque(maxlen=1024)

        self._first_snapshot: Optional[Snapshot] = None

    @property
    def incoming_history(self) -> deque:
        return self._incoming_history

    @property
    def outgoing_history(self) -> deque:
        return self._outgoing_history

    def update(self, snapshot: Snapshot) -> None:
        """喂入新的采样快照，重新计算统计"""
        self._samples.append(snapshot)
        if self._first_snapshot is None:
            self._first_snapshot = snapshot

        n = len(self._samples)
        if n < 2:
            return

        latest = self._samples[-1]

        # --- 当前速率 (最近 ~1s 的窗口) ---
        sec_idx = max(0, n - 1 - self._second_window)
        older = self._samples[sec_idx]
        dt = latest.timestamp - older.timestamp
        if dt > 0:
            cur_in = (latest.bytes_recv - older.bytes_recv) / dt
            cur_out = (latest.bytes_sent - older.bytes_sent) / dt
        else:
            cur_in = cur_out = 0.0

        self.incoming.current = max(cur_in, 0.0)
        self.outgoing.current = max(cur_out, 0.0)

        # 记录到历史（供 graph）
        self._incoming_history.appendleft(self.incoming.current)
        self._outgoing_history.appendleft(self.outgoing.current)

        # --- 平均速率 (整个窗口) ---
        oldest = self._samples[0]
        dt_all = latest.timestamp - oldest.timestamp
        if dt_all > 0:
            self.incoming.average = max(
                (latest.bytes_recv - oldest.bytes_recv) / dt_all, 0.0
            )
            self.outgoing.average = max(
                (latest.bytes_sent - oldest.bytes_sent) / dt_all, 0.0
            )

        # --- Min / Max ---
        if self.incoming.current > 0 or self.outgoing.current > 0 or n > 3:
            self.incoming.minimum = min(
                self.incoming.minimum, self.incoming.current
            )
            self.incoming.maximum = max(
                self.incoming.maximum, self.incoming.current
            )
            self.outgoing.minimum = min(
                self.outgoing.minimum, self.outgoing.current
            )
            self.outgoing.maximum = max(
                self.outgoing.maximum, self.outgoing.current
            )

        # 修正 inf
        if self.incoming.minimum == float("inf"):
            self.incoming.minimum = 0.0
        if self.outgoing.minimum == float("inf"):
            self.outgoing.minimum = 0.0

        # --- Total ---
        if self._first_snapshot:
            self.incoming.total = latest.bytes_recv
            self.outgoing.total = latest.bytes_sent


def format_speed(bytes_per_sec: float) -> str:
    """将 bytes/s 转为人类可读的 Bit/s 格式"""
    return format_speed_unit(bytes_per_sec, "bit")


def format_speed_unit(bytes_per_sec: float, unit: str = "bit") -> str:
    """根据单位选择格式化速率 (unit: 'bit' 或 'byte')"""
    if unit == "byte":
        b = bytes_per_sec
        units = [
            (1024 ** 3, "GB/s"),
            (1024 ** 2, "MB/s"),
            (1024,      "KB/s"),
            (1,         "B/s"),
        ]
        for threshold, u in units:
            if b >= threshold:
                return f"{b / threshold:.2f} {u}"
        return "0.00 B/s"
    else:
        bits = bytes_per_sec * 8
        units = [
            (1024 ** 3, "GBit/s"),
            (1024 ** 2, "MBit/s"),
            (1024,      "kBit/s"),
            (1,         "Bit/s"),
        ]
        for threshold, u in units:
            if bits >= threshold:
                return f"{bits / threshold:.2f} {u}"
        return "0.00 Bit/s"


def format_bytes(total_bytes: int) -> str:
    """将字节数转为人类可读格式"""
    b = float(total_bytes)
    units = [
        (1024 ** 3, "GByte"),
        (1024 ** 2, "MByte"),
        (1024,      "kByte"),
        (1,         "Byte"),
    ]
    for threshold, unit in units:
        if b >= threshold:
            return f"{b / threshold:.2f} {unit}"
    return "0.00 Byte"

File: rust/src/stats.rs
//! 流量统计引擎
//! 维护滑动窗口，计算 Cur / Avg / Min / Max / Ttl 五项指标。

use std::collections::VecDeque;

use crate::collector::Snapshot;

/// 某一方向（收/发）的统计结果
#[derive(Clone, Debug)]
pub struct TrafficStats {
    /// 当前速率 (bytes/s)
    pub current: f64,
    /// 平均速率 (bytes/s)
    pub average: f64,
    /// 最小速率 (bytes/s)
    pub minimum: f64,
    /// 最大速率 (bytes/s)
    pub maximum: f64,
    /// 累计字节数
    pub total: u64,
}

impl Default for TrafficStats {
    fn default() -> Self {
        Self {
            current: 0.0,
            average: 0.0,
            minimum: f64::INFINITY,
            maximum: 0.0,
            total: 0,
        }
    }
}

/// 统计引擎：为单个网卡维护收发两个方向的统计数据
pub struct StatisticsEngine {
    samples: VecDeque<Snapshot>,
    second_window: usize,
    max_samples: usize,
    sample_count: usize,

    /// 收方向统计
    pub incoming: TrafficStats,
    /// 发方向统计
    pub outgoing: TrafficStats,

    /// 收方向速率历史 (front = 最新值，供图形绘制)
    pub incoming_history: VecDeque<f64>,
    /// 发方向速率历史
    pub outgoing_history: VecDeque<f64>,
}

impl StatisticsEngine {
    pub fn new(refresh_interval_ms: u64, average_window_sec: u64) -> Self {
        let second_window = (1000u64 / refresh_interval_ms).max(1) as usize;
        let max_samples =
            ((1000u64 / refresh_interval_ms) * average_window_sec).max(600) as usize;

        Self {
            samples: VecDeque::with_capacity(max_samples),
            second_window,
            max_samples,
            sample_count: 0,
            incoming: TrafficStats::default(),
            outgoing: TrafficStats::default(),
            incoming_history: VecDeque::with_capacity(1024),
            outgoing_history: VecDeque::with_capacity(1024),
        }
    }

    /// 喂入新的采样快照，重新计算统计
    pub fn update(&mut self, snapshot: Snapshot) {
        self.samples.push_back(snapshot);
        if self.samples.len() > self.max_samples {
            self.samples.pop_front();
        }
        self.sample_count += 1;

        let n = self.samples.len();
        if n < 2 {
            return;
        }

        let latest = &self.samples[n - 1];

        // ── 当前速率 (最近 ~1s 的窗口) ──
        let sec_idx = if n > self.second_window + 1 {
            n - 1 - self.second_window
        } else {
            0
        };
        let older = &self.samples[sec_idx];
        let dt = latest.elapsed_secs - older.elapsed_secs;

        if dt > 0.0 {
            self.incoming.current =
                ((latest.bytes_recv as f64 - older.bytes_recv as f64) / dt).max(0.0);
            self.outgoing.current =
                ((latest.bytes_sent as f64 - older.bytes_sent as f64) / dt).max(0.0);
        }

        // 记录到历史 (graph 用)
        if self.incoming_history.len() >= 1024 {
            self.incoming_history.pop_back();
        }
        if self.outgoing_history.len() >= 1024 {
            self.outgoing_history.pop_back();
        }
        self.incoming_history.push_front(self.incoming.current);
        self.outgoing_history.push_front(self.outgoing.current);

        // ── 平均速率 (整个窗口) ──
        let oldest = &self.samples[0];
        let dt_all = latest.elapsed_secs - oldest.elapsed_secs;
        if dt_all > 0.0 {
            self.incoming.average =
                ((latest.bytes_recv as f64 - oldest.bytes_recv as f64) / dt_all).max(0.0);
            self.outgoing.average =
                ((latest.bytes_sent as f64 - oldest.bytes_sent as f64) / dt_all).max(0.0);
        }

        // ── Min / Max ──
        if self.incoming.current > 0.0 || self.outgoing.current > 0.0 || self.sample_count > 3 {
            self.incoming.minimum = self.incoming.minimum.min(self.incoming.current);
            self.incoming.maximum = self.incoming.maximum.max(self.incoming.current);
            self.outgoing.minimum = self.outgoing.minimum.min(self.outgoing.current);
            self.outgoing.maximum = self.outgoing.maximum.max(self.outgoing.current);
        }

        // 修正 infinity
        if self.incoming.minimum == f64::INFINITY {
            self.incoming.minimum = 0.0;
        }
        if self.outgoing.minimum == f64::INFINITY {
            self.outgoing.minimum = 0.0;
        }

        // ── Total ──
        self.incoming.total = latest.bytes_recv;
        self.outgoing.total = latest.bytes_sent;
    }
}

// ─── 格式化工具函数 ───────────────────────────────────────

use crate::Unit;

/// 根据单位选择格式化速率
pub fn format_speed_unit(bytes_per_sec: f64, unit: Unit) -> String {
    match unit {
        Unit::Bit => {
            let bits = bytes_per_sec * 8.0;
            if bits >= 1024.0 * 1024.0 * 1024.0 {
                format!("{:.2} GBit/s", bits / (1024.0 * 1024.0 * 1024.0))
            } else if bits >= 1024.0 * 1024.0 {
                format!("{:.2} MBit/s", bits / (1024.0 * 1024.0))
            } else if bits >= 1024.0 {
                format!("{:.2} kBit/s", bits / 1024.0)
            } else {
                format!("{:.2} Bit/s", bits)
            }
        }
        Unit::Byte => {
            if bytes_per_sec >= 1024.0 * 1024.0 * 1024.0 {
                format!("{:.2} GB/s", bytes_per_sec / (1024.0 * 1024.0 * 1024.0))
            } else if bytes_per_sec >= 1024.0 * 1024.0 {
                format!("{:.2} MB/s", bytes_per_sec / (1024.0 * 1024.0))
            } else if bytes_per_sec >= 1024.0 {
                format!("{:.2} KB/s", bytes_per_sec / 1024.0)
            } else {
                format!("{:.2} B/s", bytes_per_sec)
            }
        }
    }
}

/// 将字节数转为人类可读格式
pub fn format_bytes(total_bytes: u64) -> String {
    let b = total_bytes as f64;
    if b >= 1024.0 * 1024.0 * 1024.0 {
        format!("{:.2} GByte", b / (1024.0 * 1024.0 * 1024.0))
    } else if b >= 1024.0 * 1024.0 {
        format!("{:.2} MByte", b / (1024.0 * 1024.0))
    } else if b >= 1024.0 {
        format!("{:.2} kByte", b / 1024.0)
    } else {
        format!("{:.2} Byte", b)
    }
}
